### PR TITLE
Implement Textarea for Theming

### PR DESCRIPTION
**Pull request by JakeLaCombe**
### Summary
This updates the Textarea to be in it's own package, and also to be ready for OCS.

### Additional Details
New additions to textarea
- Added prop of isInvalid to indicate textarea is invalid.
- Added ability to auto resize textarea based on input.
- Added set sizes that adjusts the rows attribute of the textarea.
- Added the ability to create a textarea that is full width and height of container (Ex: Full Page Size)

Deployment link: 
https://terra-core-deployed-pr-982.herokuapp.com/static/#/site/form-textarea

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

File: packages/terra-form-textarea/package.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    "react-dom": "^15.4.2",
+    "terra-base": "^2.0.0"
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "prop-types": "^15.5.8",
+    "terra-base": "^2.0.0"
+  },
+  "scripts": {
+    "compile": "npm run compile:clean && npm run compile:build",
```
Comment (mjhenkes) [inline packages/terra-form-textarea/package.json]: You'll have to update these based on @emilyrohrbough's generator changes https://github.com/cerner/generator-terra-module/pull/71

File: package.json [hunk 2]
```diff
@@ -101,5 +101,9 @@
     "terra-toolkit": "^2.1.0",
     "webpack": "^3.6.0",
     "webpack-dev-server": "2.7.1"
+  },
```
Comment (emilyrohrbough) [inline package.json]: These should be removed

File: packages/terra-form-textarea/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+ChangeLog
+=========
+
+1.0.0 - (<!--- Insert date using June 28, 2017 format. --->)
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/CHANGELOG.md]: Update from 1.0.0 to Unreleased

File: packages/terra-form-textarea/docs/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+<Textarea
+  name="description"
+  value="I'm a software engineer working at Cerner"
+  required
+ />
+```
+
+## Component Features
+
+<!-- Uncomment supported features.
```
Comment (jmsv6d) [inline packages/terra-form-textarea/docs/README.md]: Uncomment these

File: packages/terra-form-textarea/docs/MigrationGuide.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Migrating from terra-form's textarea Component to terra-form-textarea
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/docs/MigrationGuide.md]: Add:
* examples of the old usage vs new usage.
* Add old & new prop tables.

File: packages/terra-form-textarea/docs/MigrationGuide.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/docs/MigrationGuide.md]: Additionally, lets remove this file. I just double checked with @mjhenkes & we'll include the migration guide in the release notes.

File: packages/terra-form-textarea/docs/MigrationGuide.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+- `isAutoResizable` - indicates whether the textarea can be auto-resized vertically.
+- `isInvalid` - indicates whether or not the textarea is invalid.
+- `size` - the size of the textarea.
+
+## Behavior Changes
+
+#### Added/Changed Behavior
+- `disabled` - disables the textarea if set to true.
+- `isAutoResizable` - when set to true, the textarea automatically resizes vertically when users type in the small textarea.
+- `isInvalid` - when set to true, the textarea is styles with it's error styles.
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/docs/MigrationGuide.md]: typo; the textarea is _styled_

File: packages/terra-form-textarea/docs/MigrationGuide.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+#### Props Added
+- `disabled` - Whether the textarea is disabled.
+- `isAutoResizable` - indicates whether the textarea can be auto-resized vertically.
+- `isInvalid` - indicates whether or not the textarea is invalid.
+- `size` - the size of the textarea.
+
+## Behavior Changes
+
+#### Added/Changed Behavior
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/docs/MigrationGuide.md]: IMO it makes sense to explain `isAutoreSizable` here, but it feels like the prop description covers the behavior of the other props

File: packages/terra-form-textarea/docs/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# Terra Form Textarea
+
+Element for building out Textareas in a form.
+
+**Note:** This replaces the previous implementation of textarea inside terra-form. For migration steps, consult [the migration guide](https://github.com/cerner/terra-core/tree/master/packages/terra-form-textarea/docs/MigrationGuide.md)
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/docs/README.md]: Remove

File: packages/terra-form-textarea/docs/README.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+<Textarea
+  name="description"
+  value="I'm a software engineer working at Cerner"
+  required
+ />
+```
+
+## Component Features
+
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/docs/README.md]: Uncomment supported features

File: packages/terra-form-textarea/docs/MigrationGuide.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# Migrating from terra-form's textarea Component to terra-form-textarea
+
+The following sections describe how to migrate from the textarea component provided in the terra-form package to the textarea component provided in the terra-form-textarea package. Terra-form-textarea provides enhanced API and layout to begin the first phase of the OSC uplift.
```
Comment (jmsv6d) [inline packages/terra-form-textarea/docs/MigrationGuide.md]: *OCS

I'm wondering if we should really include this in the documentation though, any consumers outside of cerner would have no idea what this means. Thoughts?

File: packages/terra-site/src/examples/form-textarea/TextareaExamples.jsx [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
```
Comment (emilyrohrbough) [inline packages/terra-site/src/examples/form-textarea/TextareaExamples.jsx]: I know this is a bit late to the game, but we should update this example to also be able to render isInvalid state. Possibly when the input length is 0.

File: packages/terra-form-textarea/package.json [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  "peerDependencies": {
+    "react": "^15.4.2",
+    "react-dom": "^15.4.2",
+    "terra-base": "^2.0.0"
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "prop-types": "^15.5.8",
+    "terra-base": "^2.0.0"
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/package.json]: "terra-base": "^2.6.1"

File: packages/terra-form-textarea/package.json [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    "url": "https://github.com/cerner/terra-core/issues"
+  },
+  "homepage": "https://github.com/cerner/terra-core#readme",
+  "devDependencies": {
+    "terra-props-table": "^1.4.0"
+  },
+  "peerDependencies": {
+    "react": "^15.4.2",
+    "react-dom": "^15.4.2",
+    "terra-base": "^2.0.0"
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/package.json]: "terra-base": "^2.6.1"

File: packages/terra-form-textarea/README.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
+# Terra Form Textarea
+
+
+[![NPM version](http://img.shields.io/npm/v/terra-form-textarea.svg)](https://www.npmjs.org/package/terra-form-textarea)
+[![Build Status](https://travis-ci.org/cerner/terra-core.svg?branch=master)](https://travis-ci.org/cerner/terra-core)
+
+Textarea component for building forms.
```
Comment (emilyrohrbough) [inline packages/terra-form-textarea/README.md]: Can we update the descriptions to match in both ReadMes & the package.json?

File: package.json (hunk 2 above)
Comment (JakeLaCombe) [reply package.json]: Sorry. Forgot I added them here.

https://github.com/cerner/terra-core/pull/982/commits/a296bdf48a4acfbc9cb55ae29a66e87ec126b55e

File: packages/terra-form-textarea/README.md (hunk 15 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/README.md]: Sure can! https://github.com/cerner/terra-core/pull/982/commits/9e911990d3df96486e60680870d9598b0f6d6579

File: packages/terra-form-textarea/docs/MigrationGuide.md (hunk 11 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/MigrationGuide.md]: I agree as well. I have removed the migration guide, so it's no longer here. https://github.com/cerner/terra-core/pull/982/commits/802777623b56e1099bdf699936a4de4e2853d6ff

In addition, I'll give the textarea a new text description. I think when we talk about OCS, we should just say the component is themeable. How does that sound?

File: packages/terra-form-textarea/docs/README.md (hunk 10 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/README.md]: Thanks!

https://github.com/cerner/terra-core/pull/982/commits/fbd4c8262dd1a7782758ec7236b6509a6ed245be

File: packages/terra-form-textarea/docs/README.md (hunk 9 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/README.md]: Done. https://github.com/cerner/terra-core/pull/982/commits/802777623b56e1099bdf699936a4de4e2853d6ff

File: packages/terra-form-textarea/docs/MigrationGuide.md (hunk 8 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/MigrationGuide.md]: Removed migration guide. https://github.com/cerner/terra-core/pull/982/commits/802777623b56e1099bdf699936a4de4e2853d6ff

File: packages/terra-form-textarea/docs/MigrationGuide.md (hunk 7 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/MigrationGuide.md]: Removed migration guide. https://github.com/cerner/terra-core/pull/982/commits/802777623b56e1099bdf699936a4de4e2853d6ff

File: packages/terra-form-textarea/docs/MigrationGuide.md (hunk 6 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/MigrationGuide.md]: Removed migration guide. https://github.com/cerner/terra-core/pull/982/commits/802777623b56e1099bdf699936a4de4e2853d6ff

File: packages/terra-form-textarea/docs/MigrationGuide.md (hunk 5 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/MigrationGuide.md]: Removed migration guide. https://github.com/cerner/terra-core/pull/982/commits/802777623b56e1099bdf699936a4de4e2853d6ff

File: packages/terra-site/src/examples/form-textarea/TextareaExamples.jsx (hunk 12 above)
Comment (JakeLaCombe) [reply packages/terra-site/src/examples/form-textarea/TextareaExamples.jsx]: Got it covered on the second one actually! https://github.com/cerner/terra-core/pull/982/files#diff-9a8da08c4bb9abc6d8eb06fa7f4dc9a9R39

Keep in mind that default doesn't have errored out styles, so you won't see them on terra-core. They are in the consumer theme though.

File: packages/terra-form-textarea/docs/README.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+## Getting Started
+
+- Install with [npmjs](https://www.npmjs.com):
+  - `npm install terra-form-textarea`
+
+## Usage
+
+```jsx
+import React from 'react';
+import Textarea from 'terra-form/lib/Textarea';
```
Comment (bjankord) [inline packages/terra-form-textarea/docs/README.md]: This should be update to `import Textarea from 'terra-form-textarea';`

File: packages/terra-form-textarea/package.json (hunk 1 above)
Comment (bjankord) [reply packages/terra-form-textarea/package.json]: Yes we need these updated based on the generator changes so this package's npm scripts will be compatible on Windows.

Comment (JakeLaCombe) [reply packages/terra-form-textarea/package.json]: Fixed here. https://github.com/cerner/terra-core/pull/982/commits/cd1b3406beecc6afa02bd2f687034c7cbf190cc5

File: packages/terra-form-textarea/package.json (hunk 14 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/package.json]: Done - https://github.com/cerner/terra-core/pull/982/commits/5b0dc5666f3d39669cc653d415a1878d42ea33af

File: packages/terra-form-textarea/package.json (hunk 13 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/package.json]: Done - https://github.com/cerner/terra-core/pull/982/commits/5b0dc5666f3d39669cc653d415a1878d42ea33af

File: packages/terra-form-textarea/docs/README.md (hunk 4 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/docs/README.md]: https://github.com/cerner/terra-core/pull/982/commits/fbd4c8262dd1a7782758ec7236b6509a6ed245be

File: packages/terra-form-textarea/CHANGELOG.md (hunk 3 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/CHANGELOG.md]: https://github.com/cerner/terra-core/commit/465155f57671f98ebe638ec1ceb0fa2962fab295

**Comment by mmalaker** [comment]
Here are a few issues I found when testing:
- We should only provide examples for functionality that is discernible to a user. The error examples don't actually show any errors.
- The disabled examples are disabled for text input but can still be resized. Resizing should also be disabled.
- Any theming changes should be viewable from the mock (or some other) theme in the dropdown list. The values don't need to conform to a specific theme; they just need to be different from the default.

File: packages/terra-form-textarea/src/Textarea.scss [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+    background-position: 0 100%;
+    background-repeat: no-repeat;
+    background-size: var(--terra-form-textarea-focus-background-start, 0 0);
+    border-color: var(--terra-form-textarea-border-color, #dedfe0);
+    border-radius: var(--terra-form-textarea-border-radius, 3px);
+    border-style: var(--terra-form-textarea-border-style, solid);
+    border-width: var(--terra-form-textarea-border-width, 1px);
+    color: var(--terra-form-textarea-color, #000);
+    cursor: var(--terra-form-textarea-cursor, auto);
+    font-family: var(--terra-base-font-family, 'Helvetica Neue', Helvetica, Arial, sans-serif);
```
Comment (bjankord) [inline packages/terra-form-textarea/src/Textarea.scss]: Why is font-family being set here? Shouldn't this be inherited from font-family being set on the [HTML element](https://github.com/cerner/terra-core/blob/master/packages/terra-base/src/Base.scss#L8)?

File: packages/terra-form-textarea/src/Textarea.scss [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+      background-color: var(--terra-form-textarea-disabled-background-color, rgba(111, 116, 119, 0.05));
+      border-color: var(--terra-form-textarea-disabled-border-color, #dedfe0);
+      border-style: var(--terra-form-textarea-disabled-border-style, solid);
+      border-width: var(--terra-form-textarea-disabled-border-width, 1px);
+      color: var(--terra-form-textarea-disabled-color, rgba(0, 0, 0, 0.25));
+      opacity: var(--terra-form-textarea-disabled-opacity, 1);
+      resize: none;
+    }
+
+    &:focus {
```
Comment (bjankord) [inline packages/terra-form-textarea/src/Textarea.scss]: Based on [this finding](https://github.com/cerner/terra-core/issues/972#issuecomment-341928670) @JakeLaCombe, I want to remove the mouse only focus styles and make the keyboard focus styles the default focus styles for this PR. Once we land a solution to determine focus source, we can re-add the different focus styles back to textarea in a future PR.

File: packages/terra-form-textarea/src/Textarea.scss (hunk 17 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/src/Textarea.scss]: What we are finding is that browsers are setting the font-family on inputs and textareas.

````
    font-family: system-ui;
````

After showing to UX, they wanted to set the value of font-family to the same value as --terra-base-font-family. Our options are to either create a new theming variable for textarea and default it to the value of terra-base-font-family, or include the base variable inside the textarea. I feel like the second approach is better as it enforces the textarea text to be the same as the base text.

Comment (bjankord) [reply packages/terra-form-textarea/src/Textarea.scss]: Let's test setting this to `font-family: inherit;` This should override the browser default styles that render text inside form elements as that of the user's operating system font while also inheriting the font-family from the HTML provided font-family isn't set anywhere on parent elements between the textarea and the HTML element.

My concern with the current code is, if we need to update hard-coded default font, we need to remember all the instances we set it as a fallback and update those. If we set font-family to inherit, it should achieve more or less the same look while allowing us maintain 1 place to change the font-family used throughout all the components.

File: packages/terra-form-textarea/src/Textarea.scss (hunk 18 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/src/Textarea.scss]: Done. https://github.com/cerner/terra-core/pull/982/commits/9893550863f5a1367263a25c8b357aa79768bb39

File: packages/terra-form-textarea/src/Textarea.scss (hunk 17 above)
Comment (JakeLaCombe) [reply packages/terra-form-textarea/src/Textarea.scss]: It works. https://github.com/cerner/terra-core/pull/982/commits/75bb0887272e5d1c74f482490b20e6d5275916b3

**Comment by mmalaker** [comment]
Functional verification completed. The text area examples display and function correctly, reflecting the theming changes.

**Comment by JakeLaCombe** [comment]
I'll look again, but the latest merge appeared to render it.

**Comment by mmalaker** [comment]
Functional verification completed. Text area displays and functions correctly with the most recent updates.

**Comment by neilpfeiffer** [comment]
Tested: PR #982
Resolves: Issue #980

**Reviewed:**
- Follows pattern considerations documented in Cerner Design Standards: [Form-Text Area](https://wiki.ucern.com/x/sQc0Yg)
- Follows UX defined design implementation for Form-Text Area
- Follows defined theme designs: terra-theme


**Fix with Enhancements: Issue #1018**
- add minimum width to stop resizing at a fixed point (specifically applicable for Chrome)
- fix scrolling error inside automatically resizing textarea after user has manually resized (and stopped the automatic resizing)


**Suggestions: Issue #1019**
- add headers above each example 
- add/reorder examples to show 1.resizable at small size (two rows), 2.fixed-small size, 3.fixed-medium size, 4.fixed-large size, 5.fill-containter (same as current), 6.textarea-small size that becomes invalid/error when empty, 7.disabled (same as current), 8.disabled with error (same as current)


Verified by OCS initial review, Ready for Release.

**Pass** |
-----------|
UX Review|

**Pull request merged**